sdks/java/core/src/main/java/org/apache/beam/sdk/io/WholeFileIO.java

```diff
@@ -0,0 +1,241 @@
+
+import org.apache.beam.sdk.io.fs.MatchResult;
+import org.apache.beam.sdk.io.fs.ResourceId;
+import org.apache.beam.sdk.options.ValueProvider;
+import org.apache.beam.sdk.transforms.Create;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.transforms.View;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.PBegin;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.PCollectionView;
+import org.apache.beam.sdk.values.PDone;
+
+/**
+ * WholeFileIO.
+ */
+public class WholeFileIO {
+
+  public static Read read() {
+    return new AutoValue_WholeFileIO_Read.Builder().build();
+  }
+
+  public static Write write() {
+    return new AutoValue_WholeFileIO_Write.Builder().build();
+  }
+
+  /**
+   * Implements read().
+   */
+  @AutoValue
+  public abstract static class Read extends PTransform<PBegin, PCollection<KV<String, byte[]>>> {
+    @Nullable
+    abstract ValueProvider<String> getFilePattern();
+
+    abstract Builder toBuilder();
+
+    @AutoValue.Builder
+    abstract static class Builder {
+      abstract Builder setFilePattern(ValueProvider<String> filePattern);
+
+      abstract Read build();
+    }
+
+    public Read from(String filePattern) {
+      checkNotNull(filePattern, "FilePattern cannot be empty.");
+      return from(ValueProvider.StaticValueProvider.of(filePattern));
+    }
+
+    public Read from(ValueProvider<String> filePattern) {
+      checkNotNull(filePattern, "FilePattern cannot be empty.");
+      return toBuilder().setFilePattern(filePattern).build();
+    }
+
+    @Override
+    public PCollection<KV<String, byte[]>> expand(PBegin input) {
+      checkNotNull(
+          getFilePattern(),
+          "Need to set the filePattern of a WholeFileIO.Read transform."
+      );
+
+      String filePattern = getFilePattern().get();
+
+      PCollection<String> filePatternPCollection = input.apply(Create.of(filePattern));
+
+      PCollection<ResourceId> resourceIds = filePatternPCollection.apply(
+          ParDo.of(
+              new DoFn<String, ResourceId>() {
+                @ProcessElement
+                public void processElement(ProcessContext c) {
+                  String filePattern = c.element();
+                  try {
+                    List<MatchResult> matchResults = FileSystems.match(
+                        Collections.singletonList(filePattern));
+                    for (MatchResult matchResult : matchResults) {
+                      List<MatchResult.Metadata> metadataList = matchResult.metadata();
+                      for (MatchResult.Metadata metadata : metadataList) {
+                        ResourceId resourceId = metadata.resourceId();
+                        c.output(resourceId);
+                      }
+                    }
+                  } catch (IOException e) {
+                    e.printStackTrace();
```

Comment:
See https://beam.apache.org/contribute/ptransform-style-guide/#runtime-errors-and-data-consistency , specifically "Never, ever, ever do this"